src/stream/src/executor/source/mod.rs

```diff
@@ -60,10 +60,22 @@ pub fn get_split_offset_mapping_from_chunk(
     offset_idx: usize,
 ) -> Option<HashMap<SplitId, String>> {
     let mut split_offset_mapping = HashMap::new();
-    for (_, row) in chunk.rows() {
-        let split_id = row.datum_at(split_idx).unwrap().into_utf8().into();
-        let offset = row.datum_at(offset_idx).unwrap().into_utf8();
-        split_offset_mapping.insert(split_id, offset.to_string());
+    if chunk.cardinality() == 0 {
+        // assumes the chunk is a heartbeat chunk
+        for i in 0..chunk.capacity() {
+            let (_, row, _) = chunk.row_at(i);
+            let split_id = row.datum_at(split_idx).unwrap().into_utf8().into();
+            let offset = row.datum_at(offset_idx).unwrap().into_utf8();
+            split_offset_mapping
+                .entry(split_id)
+                .or_insert(offset.to_string());
```

Comment:
Previously before [#14524](https://github.com/risingwavelabs/risingwave/pull/14524/files#diff-d20a701c6d209a35e84545b11ccd13ab2f332a68aa175eb4bf73edf67184e8bcL644-L648), a heartbeat message can be interleaved with data message in a message batch, use `or_insert` to prevent the offset overwritten by heartbeat. Since offset of data message is more fresher.
But in this PR we can simply use `insert`, because the heartbeat message has been extracted from the message batch and guaranteed to emit before the data chunk.